src/Altinn.App.Core/Internal/Data/CachedInstanceDataAccessor.cs

```diff
@@ -40,120 +49,314 @@ IAppModel appModel
         Instance = instance;
         _dataClient = dataClient;
         _appMetadata = appMetadata;
-        _appModel = appModel;
+        _modelSerializationService = modelSerializationService;
     }
 
     public Instance Instance { get; }
 
-    public async Task<object?> GetSingleDataByType(string dataType)
+    /// <inheritdoc />
+    public async Task<object> GetFormData(DataElementId dataElementId)
     {
-        var appMetadata = await _appMetadata.GetApplicationMetadata();
-        var dataTypeObj = appMetadata.DataTypes.Find(d => d.Id == dataType);
-        if (dataTypeObj == null)
+        return await _formDataCache.GetOrCreate(
+            dataElementId,
+            async () =>
+            {
+                var binaryData = await GetBinaryData(dataElementId);
+
+                return _modelSerializationService.DeserializeFromStorage(binaryData.Span, GetDataType(dataElementId));
+            }
+        );
+    }
+
+    public async Task<ReadOnlyMemory<byte>> GetBinaryData(DataElementId dataElementId) =>
+        await _binaryCache.GetOrCreate(
+            dataElementId,
+            async () =>
+                await _dataClient.GetDataBytes(_org, _app, _instanceOwnerPartyId, _instanceGuid, dataElementId.Guid)
+        );
+
+    /// <inheritdoc />
+    public DataElement GetDataElement(DataElementId dataElementId)
+    {
+        return Instance.Data.Find(d => d.Id == dataElementId.Id)
+            ?? throw new InvalidOperationException($"Data element with id {dataElementId.Id} not found in instance");
+    }
+
+    public DataType GetDataType(DataElementId dataElementId)
+    {
+        var dataElement = GetDataElement(dataElementId);
+        var appMetadata = _appMetadata.GetApplicationMetadata().Result;
+        var dataType = appMetadata.DataTypes.Find(d => d.Id == dataElement.DataType);
+        if (dataType is null)
         {
-            throw new InvalidOperationException($"Data type {dataType} not found in app metadata");
+            throw new InvalidOperationException($"Data type {dataElement.DataType} not found in instance");
         }
-        if (dataTypeObj.MaxCount != 1)
+
+        return dataType;
+    }
+
+    /// <inheritdoc />
+    public void AddFormDataElement(string dataTypeString, object data)
+    {
+        var dataType = GetDataTypeByString(dataTypeString).Result;
```

Comment:
Any way to avoid `.Result`? 😅

Reply:
Yes, this is somewhat bad 😊.
I'm thinking that reads to applicationmetadata should not need to be asynchronous (nor should any other file that we can cache in memory), but they currently are asynchronous for the first read of the file, which is why I think the use of `.Result` is appropriate as a temporary workaround. Would you be OK, if I make `GetDataTypeByString` synchronous with a `.Result` on the applicationmetadata read? or do you think everything should be async until we actually have a way to read the file into memory build time?